### PR TITLE
Remove singleton from structured dataset transformer engine

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
Make the `register_handler` a class method so that we can get rid of the singleton.

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [ ] Plugin
 - [x] Refactor

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Get rid of the singleton by making a few functions class methods.  This relies on our being able to isolate out the logic for stateful stuff (keeping track of registered handlers/default formats and protocols) with the non-stateful stuff (to literal and to python value and get literal type).

### CODE REVIEW

File: .github/workflows/pythonbuild.yml [hunk 1]
```diff
@@ -39,9 +39,11 @@ jobs:
         python -m pip install --upgrade pip==21.2.4 setuptools wheel
         make setup${{ matrix.spark-version-suffix }}
         pip freeze
+    - name: Test FlyteSchema compatibility
```
Comment (wild-endeavor) [inline .github/workflows/pythonbuild.yml:42]: can revert this, either way, doesn't really matter.  i think i like not running the compat stuff with coverage.

File: flytekit/types/structured/structured_dataset.py [hunk 2]
```diff
@@ -386,21 +390,22 @@ def register_handler(self, h: Handlers, default_for_type: Optional[bool] = True,
...
-            self.DEFAULT_FORMATS[h.python_type] = h.supported_format
-            self.DEFAULT_PROTOCOLS[h.python_type] = h.protocol
+            cls.DEFAULT_FORMATS[h.python_type] = h.supported_format
+            cls.DEFAULT_PROTOCOLS[h.python_type] = h.protocol
 
         # Register with the type engine as well
         # The semantics as of now are such that it doesn't matter which order these transformers are loaded in, as
         # long as the older Pandas/FlyteSchema transformer do not also specify the override
-        TypeEngine.register_additional_type(self, h.python_type, override=True)
+        engine = StructuredDatasetTransformerEngine()
```
Comment (wild-endeavor) [inline flytekit/types/structured/structured_dataset.py:407]: this is main change - every time we register a new handler, we create a new instance of this class.  not the best, but okay.  we can also cache it as a global variable (which already exists at the bottom of this .py file)... either way.

**Review by pingsutw — APPROVED** [review]

**Review by pingsutw — APPROVED** [review]

File: plugins/flytekit-papermill/dev-requirements.in [hunk 3]
```diff
@@ -1 +1,2 @@
-flytekitplugins-spark>=0.30.0b4
+git+https://github.com/flyteorg/flytekit@add-sd-make-class-methods#egg=flytekitplugins-spark&subdirectory=plugins/flytekit-spark
+# vcs+protocol://repo_url/#egg=pkg&subdirectory=flyte
```
Comment (eapolinario) [inline plugins/flytekit-papermill/dev-requirements.in:2]: why is this needed?

File: plugins/flytekit-papermill/dev-requirements.in [hunk 4]
```diff
@@ -1 +1,2 @@
-flytekitplugins-spark>=0.30.0b4
+git+https://github.com/flyteorg/flytekit@add-sd-make-class-methods#egg=flytekitplugins-spark&subdirectory=plugins/flytekit-spark
```
Comment (eapolinario) [inline plugins/flytekit-papermill/dev-requirements.in:1]: this should be moved to `master` in a follow-up PR?

Comment (wild-endeavor) [reply plugins/flytekit-papermill/dev-requirements.in:1]: yeah, i'll do as part of my next pr.
